contracts/vortex/CarbonVortex.sol

```diff
@@ -138,109 +345,508 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
+        // cache trading start time to save gas
+        uint32 tradingStartTime = _tradingStartTimes[token];
+        // revert if trading hasn't been enabled for a token
+        if (tradingStartTime == 0) {
+            revert TradingDisabled();
+        }
+        // get time elapsed since trading was enabled
+        uint32 timeElapsed = uint32(block.timestamp) - tradingStartTime;
+        // get initial price as set by resetTrading
+        Price memory price = _initialPrice[token];
+        // get the halflife for the token
+        uint32 currentPriceDecayHalfLife = token == _targetToken ? _targetTokenPriceDecayHalfLife : _priceDecayHalfLife;
+        // get the current price by adjusting the amount with the exp decay formula
+        price.sourceAmount = ExpDecayMath
+            .calcExpDecay(price.sourceAmount, timeElapsed, currentPriceDecayHalfLife)
+            .toUint128();
+        // return the price
+        return price;
     }
 
     /**
-     * @dev withdraws funds held by the contract and sends them to an account
-     *
-     * requirements:
-     *
-     * - the caller must be the admin of the contract
+     * @inheritdoc ICarbonVortex
      */
-    function withdrawFunds(
+    function trade(
         Token token,
-        address payable target,
-        uint256 amount
-    ) external validAddress(target) nonReentrant onlyAdmin {
-        if (amount == 0) {
+        uint128 targetAmount
+    ) external payable nonReentrant validToken(token) greaterThanZero(targetAmount) {
+        uint128 sourceAmount;
+        if (token == _targetToken) {
+            sourceAmount = _sellFinalTargetForTarget(targetAmount);
+        } else {
+            sourceAmount = _sellTargetTokenForToken(token, targetAmount);
+        }
+        emit TokenTraded({ caller: msg.sender, token: token, sourceAmount: sourceAmount, targetAmount: targetAmount });
+    }
+
+    function _sellTargetTokenForToken(Token token, uint128 targetAmount) private returns (uint128) {
+        uint128 sourceAmount = expectedTradeInput(token, targetAmount);
+        // revert if trade requires 0 target token
+        if (sourceAmount == 0) {
+            revert InvalidTrade();
+        }
+        // check enough target token (if target token is native) has been sent for the trade
+        if (_targetToken == NATIVE_TOKEN && msg.value < sourceAmount) {
+            revert InsufficientNativeTokenSent();
+        } else {
+            // if target token is not native, transfer the tokens to the contract
+            _targetToken.safeTransferFrom(msg.sender, address(this), sourceAmount);
+        }
+        // transfer the tokens to caller
+        token.safeTransfer(msg.sender, targetAmount);
+
+        // if the target token is native, refund any excess native token to caller
+        if (_targetToken == NATIVE_TOKEN && msg.value > sourceAmount) {
+            payable(msg.sender).sendValue(msg.value - sourceAmount);
```

Comment:
Minor: the computation of `msg.value - sourceAmount` can be unchecked.
Major: shouldn't we assert `msg.value >= sourceAmount`?

Reply:
gas savings are not worth it vs the readability imo
refunding is better to be done after user has sent the funds, again standard solidity practice

Reply:
As far as I understand, my suggestion still abides the refunding-rule that you've mentioned.